programs/ssl/ssl_client2.c

```diff
@@ -3333,10 +3333,16 @@ int main( int argc, char *argv[] )
      * In case opaque keys it's the user responsibility to keep the key valid
      * for the duration of the handshake and destroy it at the end
      */
-    if( ( opt.ecjpake_pw_opaque != DFL_ECJPAKE_PW_OPAQUE ) &&
-        ( ! mbedtls_svc_key_id_is_null( ecjpake_pw_slot ) ) )
+    if( ( opt.ecjpake_pw_opaque != DFL_ECJPAKE_PW_OPAQUE ) )
     {
-        psa_destroy_key( ecjpake_pw_slot );
+        psa_key_attributes_t check_attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+        /* Verify that the key is still valid before destroying it */
+        if( psa_get_key_attributes( ecjpake_pw_slot, &check_attributes ) ==
+                PSA_SUCCESS )
```

Comment:
Sorry, this is not what we want. This program is a testing tool (for use in `ssl-opt.sh` mostly, `compat.sh` too). The reason we want to verify that the key is still valid is that the TLS code is not supposed to destroy keys it didn't create itself. If it did, we want to detect it and exit non-zero so that the calling script (`ssl-opt.sh`) case mark the corresponding test case as failed.

In a normal program (as opposed to a test program), we would just destroy the key* and be done with it. The whole point of adding this verification is so that we get test failures if the TLS code misbehaves and starts destroying keys it doesn't own.

*Actually in a lazy normal program we could even just exit the program without bothering to destroy the key or free allocated memory. The reason we clean up properly in this program is we want to be able to run memory checkers and other resource management checks on it - not that we care about the program itself, but we want to know if there are leaks in the library, and or that we need the program to not create false positives by leaking resources itself. For PSA crypto, this is the call to `mbedtls_test_helper_is_psa_leaking()` near the end of the program. All keys created by the program must be destroyed by the program before that call. Then if anything is found, we'll know it was the library that created a key and failed to destroy it.

Reply:
Suggesting something like:
```c
        if( psa_get_key_attributes( ecjpake_pw_slot, &check_attributes ) !=
                PSA_SUCCESS )
        {
            if( ret == 0 )
                ret = 1;
            mbedtls_printf( "The EC J-PAKE password key was destroyed\n" );
        }
```
(based on other similar tests near the end of the file).

Reply:
No worries, happens. Especially in long-lived PRs when with a lot of "one last thing" changes.